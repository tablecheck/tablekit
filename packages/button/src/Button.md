Buttons are used as triggers for actions. They are used in forms, toolbars, dialog footers and as
stand-alone action triggers.

Button also exports a button-group component to make it easy to display multiple buttons together.

It's important to note that for interactivity styles (hover etc) to function correctly
either `onClick` or `href` must be defined, usually this shouldn't be an issue.

**Important!** Please be careful when using `appearance="outline"` or `appearance="primary-outline"`
, it should not be picked by default and it's use case should be carefully considered. Suggested use
cases are as follows, even if under these categories usage should be discussed with at least 1 other
person.

1. Actions that must be available but we don't want to encourage the user to click.
2. On Marketing pages for aesthetic usage and used sparingly and by design only.

```tsx
import { FieldHeight } from '@tablecheck/tablekit-theme';

<div>
  <strong>Minimum button width:</strong>
  <ol style={{ listStyleType: 'decimal', paddingLeft: '15px' }}>
    {Object.keys(FieldHeight).map((key) => (
      <li key={key}>{`${key}: ${parseInt(FieldHeight[key], 10) * 3}px`}</li>
    ))}
  </ol>
</div>;
```

## Horizontal Buttons

```tsx fullwidth
import { useState, Fragment } from 'react';
import startCase from 'lodash/startCase';
import styled from '@emotion/styled';
import { faChevronDown } from '@fortawesome/pro-solid-svg-icons/faChevronDown';
import { faEdit } from '@fortawesome/pro-solid-svg-icons/faEdit';
import { faCircle } from '@fortawesome/pro-solid-svg-icons/faCircle';
import {
  Select,
  selectDarkTheme,
  selectThemeNamespace
} from '@tablecheck/tablekit-select';
import { Typography } from '@tablecheck/tablekit-typography';
import { ItemGroup } from '@tablecheck/tablekit-item';
import { Tooltip } from '@tablecheck/tablekit-tooltip';
import { Icon } from '@tablecheck/tablekit-icon';
import {
  DARK_COLORS,
  CLASSIC_COLORS,
  ZIndex,
  ThemeProvider
} from '@tablecheck/tablekit-theme';
import {
  Checkbox,
  checkboxDarkTheme,
  checkboxThemeNamespace
} from '@tablecheck/tablekit-checkbox';
import {
  buttonDarkTheme,
  buttonThemeNamespace,
  Button,
  ButtonSize,
  ButtonAppearance,
  ButtonShape,
  ButtonColor
} from './index';

const ExamplesWrapper = styled.div`
  ${({ theme }) => `background-color: ${theme.colors.canvas};`}
  padding: 10px;
  overflow: auto;
  display: grid;
  grid-template-columns: repeat(8, 1fr);
`;

const ExamplesRow = styled.div`
  display: flex;
  grid-column-start: 1;
  grid-column-end: 9;
`;

const ExamplesHeader = styled.h4`
  ${({ theme }) => `color: ${theme.colors.text};`}
  ${Typography.Heading3}
  padding: 20px 10px 10px;
  grid-column-start: 1;
  grid-column-end: 9;
`;

const darkTheme = {
  colors: DARK_COLORS,
  [buttonThemeNamespace]: buttonDarkTheme,
  [checkboxThemeNamespace]: checkboxDarkTheme,
  [selectThemeNamespace]: selectDarkTheme
};

const classicTheme = {
  colors: CLASSIC_COLORS
};

const BtnWrapper = styled.div`
  padding: 12px;
  ${({ appearance, theme }) =>
    appearance === 'inverse' ? `background: ${theme.colors.primary}` : ''};
`;

function SunSvg() {
  return (
    <svg viewBox="0 0 302.4 302.4" width="16" height="16">
      <path
        fill="currentColor"
        d="M204.8 97.6C191.2 84 172 75.2 151.2 75.2s-40 8.4-53.6 22.4c-13.6 13.6-22.4 32.8-22.4 53.6s8.8 40 22.4 53.6c13.6 13.6 32.8 22.4 53.6 22.4s40-8.4 53.6-22.4c13.6-13.6 22.4-32.8 22.4-53.6s-8.4-40-22.4-53.6zm-14.4 92.8c-10 10-24 16-39.2 16s-29.2-6-39.2-16-16-24-16-39.2 6-29.2 16-39.2 24-16 39.2-16 29.2 6 39.2 16 16 24 16 39.2-6 29.2-16 39.2zM292 140.8h-30.8c-5.6 0-10.4 4.8-10.4 10.4 0 5.6 4.8 10.4 10.4 10.4H292c5.6 0 10.4-4.8 10.4-10.4 0-5.6-4.8-10.4-10.4-10.4zM151.2 250.8c-5.6 0-10.4 4.8-10.4 10.4V292c0 5.6 4.8 10.4 10.4 10.4 5.6 0 10.4-4.8 10.4-10.4v-30.8c0-5.6-4.8-10.4-10.4-10.4zM258 243.6l-22-22c-3.6-4-10.4-4-14.4 0s-4 10.4 0 14.4l22 22c4 4 10.4 4 14.4 0s4-10.4 0-14.4zM151.2 0c-5.6 0-10.4 4.8-10.4 10.4v30.8c0 5.6 4.8 10.4 10.4 10.4 5.6 0 10.4-4.8 10.4-10.4V10.4c0-5.6-4.8-10.4-10.4-10.4zM258.4 44.4c-4-4-10.4-4-14.4 0l-22 22c-4 4-4 10.4 0 14.4 3.6 4 10.4 4 14.4 0l22-22c4-4 4-10.4 0-14.4zM41.2 140.8H10.4c-5.6 0-10.4 4.8-10.4 10.4s4.4 10.4 10.4 10.4h30.8c5.6 0 10.4-4.8 10.4-10.4 0-5.6-4.8-10.4-10.4-10.4zM80.4 221.6c-3.6-4-10.4-4-14.4 0l-22 22c-4 4-4 10.4 0 14.4s10.4 4 14.4 0l22-22c4-4 4-10.4 0-14.4zM80.4 66.4l-22-22c-4-4-10.4-4-14.4 0s-4 10.4 0 14.4l22 22c4 4 10.4 4 14.4 0s4-10.4 0-14.4z"
      />
    </svg>
  );
}

const Btn = (props) => (
  <BtnWrapper appearance={props.appearance}>
    <Button {...props} onClick={() => {}} />
  </BtnWrapper>
);

const appearances = Object.values(ButtonAppearance).filter(
  (appearance) =>
    [ButtonAppearance.Vertical, ButtonAppearance.DisabledVertical].indexOf(
      appearance
    ) === -1
);
const shapes = Object.values(ButtonShape);
const sizes = Object.values(ButtonSize);
const colors = Object.values(ButtonColor);

const appearanceOptions = appearances.map((appearance) => ({
  label: startCase(appearance),
  value: appearance
}));

const colorOptions = colors.map((color) => ({
  label: startCase(color),
  value: color
}));

const PresentationWrapper = () => {
  const [isDarkTheme, toggleTheme] = useState(false);
  const [isLoading, setIsLoading] = useState(false);
  const [buttonAppearance, setButtonAppearance] = useState(
    ButtonAppearance.Disabled
  );
  const [buttonColor, setButtonColor] = useState(ButtonColor.Primary);

  return (
    <div>
      <ThemeProvider
        useLocalStyles
        theme={isDarkTheme ? darkTheme : classicTheme}
        isDark={isDarkTheme}
      >
        <ExamplesWrapper>
          <ExamplesRow>
            <Checkbox
              onChange={() => setIsLoading(!isLoading)}
              isChecked={isLoading}
            >
              Toggle loading state
            </Checkbox>
            <Checkbox onChange={() => toggleTheme(!isDarkTheme)}>
              Use dark theme
            </Checkbox>
          </ExamplesRow>
          <ExamplesRow>
            <ItemGroup orientation="horizontal" hasIndent>
              <ItemGroup>
                <div>Button appearance</div>
                <Select
                  options={appearanceOptions}
                  value={appearanceOptions.find(
                    ({ value }) => value === buttonAppearance
                  )}
                  menuPortalTarget={document.body}
                  onChange={({ value }) => setButtonAppearance(value)}
                  styles={{
                    menuPortal: (base) => ({ ...base, zIndex: ZIndex.Modal })
                  }}
                />
              </ItemGroup>
              <ItemGroup>
                <div>Button —Åolor</div>
                <Select
                  options={colorOptions}
                  value={colorOptions.find(
                    ({ value }) => value === buttonColor
                  )}
                  menuPortalTarget={document.body}
                  onChange={({ value }) => setButtonColor(value)}
                  styles={{
                    menuPortal: (base) => ({ ...base, zIndex: ZIndex.Modal })
                  }}
                />
              </ItemGroup>
            </ItemGroup>
          </ExamplesRow>
          <ExamplesHeader>
            Appearance {startCase(buttonAppearance)}
          </ExamplesHeader>
          {shapes.map((shape) =>
            sizes.map((size) => (
              <Fragment key={`${shape}_${size}`}>
                <Btn
                  shape={shape}
                  isLoading={isLoading}
                  shouldFitContainer
                  size={size}
                  appearance={buttonAppearance}
                  color={buttonColor}
                >
                  {`${startCase(size)} ${startCase(shape)}`}
                </Btn>
                <Btn
                  shape={shape}
                  isLoading={isLoading}
                  size={size}
                  appearance={buttonAppearance}
                  color={buttonColor}
                  iconBefore={<Icon icon={faEdit} />}
                />
              </Fragment>
            ))
          )}
          <ExamplesRow>
            <Btn
              iconBefore={<Icon icon={faEdit} />}
              isLoading={isLoading}
              size={ButtonSize.Small}
            >
              Icon before Small
            </Btn>
            <Btn iconBefore={<Icon icon={faEdit} />} isLoading={isLoading}>
              Icon before
            </Btn>
            <Btn
              appearance="default"
              iconAfter={<Icon icon={faChevronDown} />}
              isLoading={isLoading}
              isDisabled
            >
              Icon after
            </Btn>
            <Btn
              iconBefore={<Icon icon={faEdit} size="large" />}
              isLoading={isLoading}
              size={ButtonSize.Large}
            >
              Icon before Large
            </Btn>
          </ExamplesRow>
          <ExamplesRow>
            <Btn
              size={ButtonSize.Small}
              shape={ButtonShape.Circular}
              iconBefore={<Icon icon={faCircle} />}
              isLoading={isLoading}
            />
            <Btn
              shape={ButtonShape.Circular}
              iconBefore={<Icon icon={faCircle} />}
              isLoading={isLoading}
            />
            <Btn
              size={ButtonSize.Large}
              shape={ButtonShape.Circular}
              iconBefore={<Icon icon={faCircle} />}
              isLoading={isLoading}
            />
          </ExamplesRow>
          <ExamplesRow>
            <Btn isDisabled href="http://google.com" isLoading={isLoading}>
              Disabled Link to Google
            </Btn>
            <Btn href="http://google.com" isLoading={isLoading}>
              Link to Google
            </Btn>
            <Btn iconBefore={SunSvg}>With Custom Icon</Btn>
            <Tooltip
              shouldHideOnMouseDown
              content="Some content"
              popperProps={{ placement: 'bottom' }}
            >
              {(setTargetRef) => (
                <Button ref={setTargetRef}>Forwarded Ref</Button>
              )}
            </Tooltip>
          </ExamplesRow>
          <ExamplesRow>
            <Btn
              shape={ButtonShape.Circular}
              appearance={ButtonAppearance.Subtle}
              iconBefore={<Icon icon={faEdit} />}
              isLoading={isLoading}
            />
            <div style={{ padding: 12 }}>
              <Button
                onClick={() => {}}
                isLoading={isLoading}
                size={{
                  'min-width: 800px': ButtonSize.Large,
                  default: ButtonSize.Regular
                }}
              >
                Size can also be responsive
              </Button>
            </div>
          </ExamplesRow>
          <ExamplesRow>
            <Button
              iconBefore={<Icon icon={faEdit} />}
              iconAfter={<Icon icon={faEdit} />}
              shouldFitContainer
              isLoading={isLoading}
            >
              Center aligned icons
            </Button>
          </ExamplesRow>
        </ExamplesWrapper>
      </ThemeProvider>
    </div>
  );
};

<PresentationWrapper />;
```

## Vertical Buttons

```tsx fullwidth
import { useState, Fragment } from 'react';
import startCase from 'lodash/startCase';
import { ThemeProvider as EmotionThemeProvider } from '@emotion/react';
import styled from '@emotion/styled';
import { faEdit } from '@fortawesome/pro-solid-svg-icons/faEdit';
import {
  Select,
  selectDarkTheme,
  selectThemeNamespace
} from '@tablecheck/tablekit-select';
import { ItemGroup } from '@tablecheck/tablekit-item';
import { Icon } from '@tablecheck/tablekit-icon';
import {
  COLORS,
  CLASSIC_COLORS,
  DARK_COLORS,
  ZIndex,
  ThemeProvider
} from '@tablecheck/tablekit-theme';
import {
  Checkbox,
  checkboxDarkTheme,
  checkboxThemeNamespace
} from '@tablecheck/tablekit-checkbox';
import {
  Button,
  ButtonSize,
  ButtonColor,
  buttonThemeNamespace,
  buttonDarkTheme,
  ButtonShape,
  ButtonAppearance
} from './index';

const ExamplesWrapper = styled.div`
  ${({ theme }) => `background-color: ${theme.colors.canvas};`}
  padding: 10px;
  overflow: auto;
  display: grid;
  grid-template-columns: repeat(4, 1fr);
`;

const ExamplesRow = styled.div`
  display: flex;
  grid-column-start: 1;
  grid-column-end: 5;
`;

const darkTheme = {
  colors: DARK_COLORS,
  [buttonThemeNamespace]: buttonDarkTheme,
  [checkboxThemeNamespace]: checkboxDarkTheme,
  [selectThemeNamespace]: selectDarkTheme
};

const classicTheme = {
  colors: CLASSIC_COLORS
};

const BtnWrapper = styled.div`
  padding: 12px;
  ${({ appearance, theme }) =>
    appearance === 'inverse' ? `background: ${theme.colors.primary}` : ''};
`;

const VerticalButton = (props) => (
  <BtnWrapper appearance={props.appearance}>
    <Button {...props} onClick={() => {}} />
  </BtnWrapper>
);

const shapes = Object.values(ButtonShape).filter(
  (shape) => shape === ButtonShape.Sharp
);
const sizes = Object.values(ButtonSize);
const colors = Object.values(ButtonColor);

const appearanceOptions = [
  {
    label: startCase(ButtonAppearance.DisabledVertical),
    value: ButtonAppearance.DisabledVertical
  },
  {
    label: startCase(ButtonAppearance.Vertical),
    value: ButtonAppearance.Vertical
  }
];
const colorOptions = colors.map((color) => ({
  label: startCase(color),
  value: color
}));

const PresentationWrapper = () => {
  const [isDarkTheme, toggleTheme] = useState(false);
  const [isLoading, setIsLoading] = useState(false);
  const [buttonAppearance, setButtonAppearance] = useState(
    ButtonAppearance.Vertical
  );
  const [buttonColor, setButtonColor] = useState(ButtonColor.Primary);

  return (
    <div>
      <ThemeProvider
        useLocalStyles
        theme={isDarkTheme ? darkTheme : classicTheme}
        isDark={isDarkTheme}
      >
        <ExamplesWrapper>
          <ExamplesRow>
            <Checkbox
              onChange={() => setIsLoading(!isLoading)}
              isChecked={isLoading}
            >
              Toggle loading state
            </Checkbox>
            <Checkbox onChange={() => toggleTheme(!isDarkTheme)}>
              Use dark theme
            </Checkbox>
          </ExamplesRow>
          <ExamplesRow>
            <ItemGroup orientation="horizontal" hasIndent>
              <ItemGroup>
                <div>Button appearance</div>
                <Select
                  options={appearanceOptions}
                  value={appearanceOptions.find(
                    ({ value }) => value === buttonAppearance
                  )}
                  menuPortalTarget={document.body}
                  onChange={({ value }) => setButtonAppearance(value)}
                  styles={{
                    menuPortal: (base) => ({ ...base, zIndex: ZIndex.Modal })
                  }}
                />
              </ItemGroup>
              <ItemGroup>
                <div>Button —Åolor</div>
                <Select
                  options={colorOptions}
                  value={colorOptions.find(
                    ({ value }) => value === buttonColor
                  )}
                  menuPortalTarget={document.body}
                  onChange={({ value }) => setButtonColor(value)}
                  styles={{
                    menuPortal: (base) => ({ ...base, zIndex: ZIndex.Modal })
                  }}
                />
              </ItemGroup>
            </ItemGroup>
          </ExamplesRow>
          {shapes.map((shape) =>
            sizes.map((size) => (
              <VerticalButton
                key={`${shape}_${size}`}
                iconBefore={<Icon icon={faEdit} />}
                isLoading={isLoading}
                size={size}
                appearance={buttonAppearance}
                color={buttonColor}
                shape={shape}
              >
                {`${startCase(size)} ${startCase(shape)}`}
              </VerticalButton>
            ))
          )}
        </ExamplesWrapper>
      </ThemeProvider>
    </div>
  );
};

<PresentationWrapper />;
```

## Button custom theming

```tsx
import { ThemeProvider as EmotionThemeProvider } from '@emotion/react';
import { faEdit } from '@fortawesome/pro-solid-svg-icons/faEdit';
import { Icon } from '@tablecheck/tablekit-icon';
import { ItemGroup } from '@tablecheck/tablekit-item';
import {
  COLORS,
  CLASSIC_COLORS,
  ThemeProvider
} from '@tablecheck/tablekit-theme';
import {
  Button,
  buttonThemeNamespace,
  ButtonAppearance,
  ButtonShape
} from './index';

const appearances = Object.values(ButtonAppearance);

const customTheme = {
  colors: {
    ...CLASSIC_COLORS,
    primary: COLORS.BLUE.L5,
    primary2: COLORS.BLUE.L7,
    danger: COLORS.PINK.L5,
    success: COLORS.AQUA.L5,
    warning: COLORS.ORANGE.L5,
    focusOutline: COLORS.ORANGE.L4
  }
};

const PresentationWrapper = () => {
  return (
    <EmotionThemeProvider theme={customTheme}>
      <ItemGroup orientation="horizontal" hasIndent>
        {appearances.map((appearance) => (
          <Button
            shape={ButtonShape.Circular}
            key={appearance}
            appearance={appearance}
            iconBefore={<Icon icon={faEdit} />}
          />
        ))}
      </ItemGroup>
    </EmotionThemeProvider>
  );
};

<ThemeProvider useLocalStyles>
  <PresentationWrapper />
</ThemeProvider>;
```

## ComponentButton usage

```tsx
import styled from '@emotion/styled';
import { ThemeProvider } from '@tablecheck/tablekit-theme';
import { useState } from 'react';
import { faChevronDown } from '@fortawesome/pro-solid-svg-icons/faChevronDown';
import { Checkbox } from '@tablecheck/tablekit-checkbox';
import { Icon } from '@tablecheck/tablekit-icon';
import { ComponentButton } from './index';

const Chevron = styled(Icon)`
  transition: 0.25s ease-in-out;
  transform: rotate(0deg);
`;
const RotatorButton = styled.button`
  &[data-state='open'] svg {
    transform: rotate(180deg);
  }
`;

const Button = ComponentButton.withComponent(RotatorButton);

function DemoComponent() {
  const [state, set] = useState('closed');
  return (
    <ThemeProvider>
      <Checkbox
        onChange={() => set(state === 'open' ? 'closed' : 'open')}
        isChecked={state === 'open'}
      >
        Toggle chevron state
      </Checkbox>
      <Button data-state={state} iconAfter={<Chevron icon={faChevronDown} />}>
        Animated Button Icon
      </Button>
    </ThemeProvider>
  );
}

<DemoComponent />;
```
