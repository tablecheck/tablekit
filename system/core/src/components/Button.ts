import { css } from '@emotion/react';

import {
  beforeStyles as spinnerBeforeStyles,
  baseStyles as spinnerElementStyles
} from './Spinner';

export const element = 'button';
export const selectors = ['button', 'a[role="button"]'];
export const className = 'btn';

export interface Props {
  /**
   * @deprecated The tooltip styling clashes with the busy styling, apply the tooltip on a wrapper component
   */
  'data-tooltip'?: never;
  'data-variant'?: ButtonVariant;
  'data-size'?: 'small' | 'large';
  'aria-busy'?: boolean;
}

const variants = ['primary', 'secondary', 'tertiary', 'ghost', 'bare'] as const;

export type ButtonVariant = (typeof variants)[number];

export const variantStyles = {
  primary: css`
    --color: white;
    --background-color: var(--primary);
    --border-color: var(--primary);
    &[data-pseudo='hover'],
    &:hover {
      --background-color: var(--primary-hover);
      --border-color: var(--primary-hover);
    }
    &[data-pseudo='active'],
    &:active {
      --background-color: var(--primary-active);
      --border-color: var(--primary-active);
    }
  `,
  secondary: css`
    --color: var(--text-contrast);
    --background-color: var(--secondary);
    --border-color: var(--secondary);
    &[data-pseudo='hover'],
    &:hover {
      --background-color: var(--secondary-hover);
      --border-color: var(--secondary-hover);
    }
    &[data-pseudo='active'],
    &:active {
      --background-color: var(--secondary-active);
      --border-color: var(--secondary-active);
    }
  `,
  tertiary: css`
    --color: var(--text);
    --background-color: var(--surface);
    --border-color: var(--border-transparent);
    &[data-pseudo='hover'],
    &:hover {
      --background-color: var(--surface-hover);
    }
    &[data-pseudo='active'],
    &:active {
      --background-color: var(--surface-active);
    }
  `,
  ghost: css`
    --color: var(--text);
    --background-color: transparent;
    --border-color: var(--border-transparent);
    &[data-pseudo='hover'],
    &:hover {
      --background-color: var(--surface-hover-transparent);
    }
    &[data-pseudo='active'],
    &:active {
      --background-color: var(--surface-active);
    }
  `,
  bare: css`
    --color: var(--text);
    --background-color: transparent;
    --border-color: transparent;
    &[data-pseudo='hover'],
    &:hover {
      --background-color: var(--surface-hover);
    }
    &[data-pseudo='active'],
    &:active {
      --background-color: var(--surface-active);
    }
  `
};

export const coreStyles = css`
  position: relative;
  display: grid;
  padding: 9px var(--spacing-l3);
  grid-gap: var(--spacing-l2);
  grid-auto-flow: column;
  white-space: nowrap;
  cursor: pointer;

  font-weight: 500;
  font-size: 16px;
  line-height: 20px;
  border: 1px solid var(--primary);
  border-radius: var(--border-radius-small);

  align-items: center;
  text-align: center;

  &[data-pseudo='focus'],
  &:focus:not(:focus-visible),
  &:focus-visible {
    outline: none;
    &:after {
      content: '';
      position: absolute;
      top: -3px;
      bottom: -3px;
      left: -3px;
      right: -3px;
      border-radius: 6px;
      border: 2px solid var(--focus, hsla(219, 78.5%, 52.5%, 1));
    }
  }

  &:disabled:disabled {
    &,
    &[data-pseudo='hover'],
    &[data-pseudo='active'],
    &:hover,
    &:active {
      cursor: not-allowed;
      pointer-events: none;
      --background-color: var(--surface-disabled);
      border-color: var(--surface-disabled);
      --color: var(--text-disabled);
    }
  }

  color: var(--color);
  background-color: var(--background-color);
  border-color: var(--border-color);
  --loading-transition: 300ms ease-in-out;
  transition: color var(--loading-transition),
    background-color var(--loading-transition),
    border-color var(--loading-transition);

  &:before {
    --spinner-size: 20px;
    ${spinnerBeforeStyles};
    display: block;
    position: absolute;
    top: calc(50% - var(--spinner-size) / 2);
    left: calc(50% - var(--spinner-size) / 2);
    opacity: 0;
    background-color: var(--color);
    transition: opacity var(--loading-transition);
    transition-delay: 0ms;
    z-index: 2;
  }
  &[aria-busy='true'] {
    ${spinnerElementStyles};
    color: transparent;
  }
  &[aria-busy='true']:before {
    opacity: 1;
    pointer-events: none;
    transition-duration: 150ms;
    transition-delay: 200ms;
  }

  &[data-size='small'] {
    padding: 7px var(--spacing-l2);
    font-size: 14px;
    line-height: 16px;
  }

  &[data-size='large'] {
    padding: 13px var(--spacing-l4);
    font-size: 16px;
    line-height: 20px;
  }
`;

export const baseStyles = css`
  ${coreStyles};

  &:not([data-variant]) {
    ${variantStyles.primary}
  }
`;
