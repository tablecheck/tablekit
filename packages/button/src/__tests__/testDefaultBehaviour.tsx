import { mount } from '@tablecheck/tablekit-enzyme';

import { Button, AnchorButton } from '../Button';

// TODO this needs to be split into AnchorButton, Button and probably ComponentButton

describe('@tablecheck/tablekit-button - default-behaviour', () => {
  test('button should have type="button" by default', () => {
    expect(
      mount(<Button />)
        .find('button')
        .props().type
    ).toBe('button');
  });

  test('should render button if there is no href property', () => {
    const wrapper = mount(<Button />);
    expect(wrapper.find('button')).toHaveLength(1);
    expect(wrapper.find('a')).toHaveLength(0);
  });

  test('should render link if href property is set', () => {
    const wrapper = mount(<AnchorButton href="test" />);
    expect(wrapper.find('a')).toHaveLength(1);
    expect(wrapper.find('button')).toHaveLength(0);
  });

  test('should render a button when the element is disabled and has href property', () => {
    const wrapper = mount(<AnchorButton isDisabled href="test" />);
    expect(wrapper.find('StyledLink')).toHaveLength(1);
    // expect(wrapper.find('a')).toHaveLength(0);
  });

  test("should not render span when the button is disabled, but doesn't have href", () => {
    const wrapper = mount(<Button isDisabled />);
    expect(wrapper.find('StyledSpan')).toHaveLength(0);
    expect(wrapper.find('button')).toHaveLength(1);
    expect(wrapper.find('a')).toHaveLength(0);
  });

  test('should render icon if the prop iconBefore is set', () => {
    const Icon = <div id="icon" />;
    const wrapper = mount(<AnchorButton href="test" iconBefore={Icon} />);
    expect(wrapper.find('#icon')).toHaveLength(1);
  });

  test('should render iconBefore before children', () => {
    const Icon = <div id="icon">icon</div>;
    const wrapper = mount(
      <AnchorButton href="test" iconBefore={Icon}>
        button
      </AnchorButton>
    );
    expect(wrapper.text()).toBe('iconbutton');
  });

  test('should render icon if the prop iconAfter is set', () => {
    const Icon = <div id="icon" />;
    const wrapper = mount(<AnchorButton href="test" iconAfter={Icon} />);
    expect(wrapper.find('#icon')).toHaveLength(1);
  });

  test('should render iconAfter after children', () => {
    const Icon = <div id="icon">icon</div>;
    const wrapper = mount(
      <AnchorButton href="test" iconAfter={Icon}>
        button
      </AnchorButton>
    );
    expect(wrapper.text()).toBe('buttonicon');
  });

  test('should render button with full container width', () => {
    const wrapper = mount(<Button shouldFitContainer />);
    expect(wrapper.find('ButtonWrapper').prop('fit')).toBe(true);
  });

  test('should render button without full container width', () => {
    const wrapper = mount(<Button />);
    expect(wrapper.find('ButtonWrapper').prop('fit')).toBe(false);
  });

  test('should be able to render both of the icons', () => {
    const Icon1 = <div id="icon1">icon1</div>;
    const Icon2 = <div id="icon2">icon2</div>;
    const wrapper = mount(
      <AnchorButton href="test" iconBefore={Icon1} iconAfter={Icon2}>
        button
      </AnchorButton>
    );
    expect(wrapper.find('#icon1')).toHaveLength(1);
    expect(wrapper.find('#icon2')).toHaveLength(1);
    expect(wrapper.text()).toBe('icon1buttonicon2');
  });

  test('should call onClick handler when link is clicked', () => {
    const spy = jest.fn();
    const wrapper = mount(
      <AnchorButton href="test" onClick={spy}>
        button
      </AnchorButton>
    );
    wrapper.find('a').simulate('click');
    expect(spy).toHaveBeenCalledTimes(1);
  });

  test('should call onClick handler when button is clicked', () => {
    const spy = jest.fn();
    const wrapper = mount(<Button onClick={spy}>button</Button>);
    wrapper.find('button').simulate('click');
    expect(spy).toHaveBeenCalledTimes(1);
  });

  test('should render tabIndex attribute when the tabIndex property is set', () => {
    let wrapper = mount(<Button tabIndex={0}>button</Button>);
    expect(wrapper.find('button').is('[tabIndex=0]')).toBe(true);
    wrapper = mount(
      <AnchorButton href="#" tabIndex={0}>
        link
      </AnchorButton>
    );
    expect(wrapper.find('a').is('[tabIndex=0]')).toBe(true);
    wrapper = mount(
      <Button tabIndex={0} isDisabled>
        span
      </Button>
    );
    expect(wrapper.find('button').is('[tabIndex=0]')).toBe(true);
  });

  describe('isLoading', () => {
    test('should render the loading spinner when isLoading is true', () => {
      const wrapper = mount(<Button isLoading>Some text</Button>);
      expect(wrapper.find('Spinner')).toHaveLength(1);
    });
    test('should not render the loading spinner when isLoading is false', () => {
      const wrapper = mount(<Button>Some text</Button>);
      expect(wrapper.find('Spinner')).toHaveLength(0);
    });
    test('should set the opacity of the text to 0 when isLoading is true', () => {
      const wrapper = mount(<Button isLoading>Some text</Button>);
      expect(wrapper.find('ButtonContent').find('span')).toMatchSnapshot();
    });

    const Icon = () => <span className="icon" />;

    test('should set the iconBefore opacity to 0 when isLoading', () => {
      const wrapper = mount(
        <Button isLoading iconBefore={<Icon />}>
          Some text
        </Button>
      );
      expect(wrapper.find('IconWrapper').find('span')).toMatchSnapshot();
    });
    test('should set the iconAfter opacity to 0 when isLoading', () => {
      const wrapper = mount(
        <Button isLoading iconAfter={<Icon />}>
          Some text
        </Button>
      );
      expect(wrapper.find('IconWrapper').find('span')).toMatchSnapshot();
    });
    test('should set the opacity of the text to undefined when isLoading is false', () => {
      const wrapper = mount(<Button>Some text</Button>);
      expect(wrapper.find('ButtonContent').find('span')).toMatchSnapshot();
    });
  });

  test('should set accessibility attributes', () => {
    expect(mount(<Button />).find('button[aria-haspopup]')).toHaveLength(0);
    expect(mount(<Button />).find('button[aria-expanded]')).toHaveLength(0);
    expect(mount(<Button />).find('button[aria-controls]')).toHaveLength(0);
    expect(mount(<Button />).find('button[id]')).toHaveLength(0);
    expect(
      mount(<Button aria-haspopup />).find('button[aria-haspopup=true]')
    ).toHaveLength(1);
    expect(
      mount(<Button aria-expanded />).find('button[aria-expanded=true]')
    ).toHaveLength(1);
    expect(
      mount(<Button aria-controls="test" />).find(
        'button[aria-controls="test"]'
      )
    ).toHaveLength(1);
    expect(mount(<Button id="test" />).find('button[id="test"]')).toHaveLength(
      1
    );
  });
});
