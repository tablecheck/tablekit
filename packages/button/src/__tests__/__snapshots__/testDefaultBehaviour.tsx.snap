// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@tablecheck/tablekit-button - default-behaviour isLoading should set the iconAfter opacity to 0 when isLoading 1`] = `
Array [
  .emotion-0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  width: auto;
  height: auto;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  margin: 0;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-transition: opacity 0.3s;
  transition: opacity 0.3s;
  opacity: 0;
}

<span
    className="emotion-0 emotion-1"
  >
    <Icon
      size="regular"
    >
      <span
        className="icon"
      />
    </Icon>
  </span>,
  <span
    className="icon"
  />,
]
`;

exports[`@tablecheck/tablekit-button - default-behaviour isLoading should set the iconBefore opacity to 0 when isLoading 1`] = `
Array [
  .emotion-0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  width: auto;
  height: auto;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  margin: 0;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-transition: opacity 0.3s;
  transition: opacity 0.3s;
  opacity: 0;
}

<span
    className="emotion-0 emotion-1"
  >
    <Icon
      size="regular"
    >
      <span
        className="icon"
      />
    </Icon>
  </span>,
  <span
    className="icon"
  />,
]
`;

exports[`@tablecheck/tablekit-button - default-behaviour isLoading should set the opacity of the text to 0 when isLoading is true 1`] = `
.emotion-0 {
  margin: 0 0 0 0;
  max-width: 100%;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  font-weight: 600;
  -webkit-transition: opacity 0.3s;
  transition: opacity 0.3s;
  opacity: 0;
}

<span
  className="emotion-0 emotion-1"
>
  Some text
</span>
`;

exports[`@tablecheck/tablekit-button - default-behaviour isLoading should set the opacity of the text to undefined when isLoading is false 1`] = `
.emotion-0 {
  margin: 0 0 0 0;
  max-width: 100%;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  font-weight: 600;
  -webkit-transition: opacity 0.3s;
  transition: opacity 0.3s;
}

<span
  className="emotion-0 emotion-1"
>
  Some text
</span>
`;
